File: modules/phish-catcher/src/intel/jcs.ts
/**
 * JCS (JSON Canonicalization Scheme, RFC 8785) — the platform's single
 * canonicalization (doc 01 §10.2; doc 07 §12: "Do not invent a second
 * canonicalization"). Same `canonicalize` package the platform TS SDK uses.
 */
import canonicalize from "canonicalize";

/** Serialize a JSON value in JCS (RFC 8785) canonical form. */
export function jcs(value: unknown): string {
  const out = canonicalize(value as never);
  if (out === undefined) throw new Error("value is not JCS-serializable");
  return out;
}

File: services/gatekeeper/internal/jsonx/jcs.go
// Package jsonx provides JCS (RFC 8785) canonical JSON serialization.
// RoE signatures, audit hash chains, and scope-manifest hashes are all
// computed over JCS-canonical bytes (doc 01 §10.2, doc 11 §3.4, Ruling A.3).
package jsonx

import (
	"encoding/json"
	"fmt"

	jsoncanonicalizer "github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
)

// Canonical marshals v to JSON and returns its JCS (RFC 8785) canonical form.
func Canonical(v any) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, fmt.Errorf("jsonx: marshal: %w", err)
	}
	canon, err := jsoncanonicalizer.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("jsonx: canonicalize: %w", err)
	}
	return canon, nil
}

// CanonicalRaw canonicalizes already-encoded JSON.
func CanonicalRaw(raw []byte) ([]byte, error) {
	canon, err := jsoncanonicalizer.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("jsonx: canonicalize: %w", err)
	}
	return canon, nil
}
